apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/CustomIdGenerator.kt

```diff
@@ -0,0 +1,24 @@
+package com.apollographql.apollo.compiler
+
+interface CustomIdGenerator {
+  fun apply(
+      /**
+       * The minified query string source with fragments
+       */
+      queryString: String,
+
+      /**
+       * The path of the query file
+       */
+      queryFilepath: String
+  ): String
+
+  /**
+   * The version of the CustomIdGenerator
+   *
+   * You should change the version every time the implementation of the CustomIdGenerator
+   * changes to let gradle and build tools know that they have to re-generate the
+   * resulting files.
+   */
+  val version: String
+}
```

Comment:
Another suggestion, what do you think of adding default implementation here:
```
interface OperationIdGenerator {
...
  object Sha256: OperationIdGenerator { ... }
  object Md5: OperationIdGenerator { ... }
}
```

So on user side it will look like this:
```
apollo {
  operationIdGenerator = OperationIdGenerator.Md5
}
```

Reply:
Instead of an `object`, I used a class: In the Kotlin plugin-incubating, I'm able to use the default value during arguments construction like so 

```
Arguments(operationIdGenerator = OperationIdGenerator.Sha256)
```

But in the Groovy version, Kotlin `object`s are `java.lang.Class`es and are not castable/assignable to class `OperationIdGenerator`. I get this error
  + `Cannot cast object 'class com.apollographql.apollo.compiler.OperationIdGenerator$Sha256' with class 'java.lang.Class' to class 'com.apollographql.apollo.compiler.OperationIdGenerator'`

Reply:
@boopathi you should be able to reference a kotlin object using `.INSTANCE`. Something like this: 

```
OperationIdGenerator gen = OperationIdGenerator.Sha256.INSTANCE
```

Reply:
That being said, it makes the groovy experience a bit weird so I like better keeping the class and not using object there.